### PR TITLE
Download revision based on const, unzip

### PR DESCRIPTION
**Pull request by leshow**
#8 

So, there's a few things to consider here: 

- The ticket said "download latest chromium". However, while I have a mechanism to get the latest revision, I currently am just using hardcoded revision. I thought it may be troublesome to dl the latest revision *always* because if we don't update the library for a while it could cause errors.

- Launching a downloaded revision requires sandboxing being enabled (https://github.com/GoogleChrome/puppeteer/blob/master/docs/troubleshooting.md) so we'd need to add something like this to the docs. If you try this branch without enabling sandboxing, chrome will fail to launch.

- The ticket you linked in the issue mentioned doing this download as part of `build.rs`. I'm not doing that, however, we could. Personally, I think it should happen when the program is launched not at compile time. Thoughts?

- I added support for mac/windows but I haven't tested.

- I've never used `failure` before, and it's entirely possible I haven't done errors in a way that's consistent with the project.

lastly:
- I've allowed `LaunchOptions` to take a user defined `path` still that will prevent any chrome revision from being downloaded. It's possible we don't want this or user defined `revisions`, and we want to strictly download a compatible revision and nothing else.

Please feel free to critique anything else, I'm happy to change anything

### CODE REVIEW

**Comment by atroche** [comment]
Sorry for slow response! Reading now.

I agree about downloading at runtime rather than in `build.rs`, definitely.

Also I think the hard-coded known version is a good solution for now at least.

**Comment by leshow** [comment]
Cool! I'll remove the code that grabs the latest revision code to start then, it's being marked as unused and making the build fail.

**Comment by atroche** [comment]
First thing I noticed is that the unzipping of the chrome-linux.zip file took about 1 minute and 15 seconds, but when I timed `unzip` on the same file, it took less than 2 seconds.

**Comment by leshow** [comment]
I noticed the same thing, but if you use `cargo run` like I did I'm pretty sure it's running the debug version. You'd need to make a build and run with --release

**Comment by atroche** [comment]
Will downstream users using this crate to in their tests get the release version, or debug? And what kind of timing did you get with release?

**Comment by atroche** [comment]
I'll test it out on my MacBook too

File: src/browser/fetcher.rs [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+    path::{Path, PathBuf},
+    str::FromStr,
+};
+
+pub const CUR_REV: &str = "634997";
+
+const APP_NAME: &str = "headless-chrome";
+const DEFAULT_HOST: &str = "https://storage.googleapis.com";
+
+const REV_URL: &str =
```
Comment (atroche) [inline src/browser/fetcher.rs]: I think you forgot a `cfg(target_os="")` here, which causes `REV_URL` to be defined twice on Mac / Windows.

Comment (atroche) [reply src/browser/fetcher.rs]: Same with PLATFORM

**Comment by leshow** [comment]
> Will downstream users using this crate to in their tests get the release version, or debug? And what kind of timing did you get with release

The version published on crates is a release version always, so users of this project won't ever run a debug version. 

AFAIK I'm using the most popular zip crate, and if you look at the code I'm not doing anything fancy. Anyway, with `cargo run --release` I'm getting around 20 seconds here. It is a little on the slow side, I'm open to using a different crate if you'd like. We could also call out to unzip or similar, but it seems pretty platform dependent and would be a shame to tie ourself to a 3rd party tool

**Comment by atroche** [comment]
Cool, no worries, 20 extra seconds for downstream users isn't the end of the world!

**Comment by leshow** [comment]
It's also a one time thing, there's that to consider

**Comment by atroche** [comment]
Getting an interesting error on my MacBook, even though it compiles fine on Linux and they're both using rustc 1.32.0:

image

**Comment by atroche** [comment]
And when I say interesting I mean I'm not sure how it's happening.

**Comment by atroche** [comment]
Ohh, I think `cfg(unix)` must match MacOS as well as Linux.

**Comment by leshow** [comment]
That makes sense, what's the proper platform for linux only?

**Comment by atroche** [comment]
`"linux"`

( https://doc.rust-lang.org/reference/conditional-compilation.html#target_os )

**Comment by atroche** [comment]
So other than getting it passing on Travis, the only other thing I'd want before we merge is something in the docs to say that if you don't supply Path as a launch option we'll download a recent Chromium version and use that.

I've got a feeling that most people aren't going to have trouble with `--no-sandbox`, so I don't mind leaving that out for now.

**Comment by leshow** [comment]
I don't think we pass `--no-sandbox` right now, so we'd need to add that if that's the route you want to go. Without that, the kernel config enabled, or the setuid sandbox, headless_chrome will just die on retries (it did this on one of my machines before I figured out the sandbox bit).

**Comment by atroche** [comment]
Oh, okay, it must be more likely that I assumed :) I reckon we could get away with a brief note and a link to the puppeteer troubleshooting page you mentioned.

And at some point we should probably give the option of no sandbox, or at least of supplying arbitrary arguments to the Chromium binary.

P.S., if you'd rather not write the docs, I'm happy to.

**Comment by atroche** [comment]
Damn, so on Travis we're getting:

`
The job exceeded the maximum log length, and has been terminated.
`

We could change the test set up code to check whether we're in CI and, if so, hard-code the path to the Chrome binary? Or we could change the log level / disable the progress bar... but it takes so long that we'd probably also want to add caching in, if we went that way.

**Comment by leshow** [comment]
Yeah I just noticed that. I'm not quite sure how to detect whether we're in a CI env or not but I'll do a big of digging.

**Comment by lukaslueg** [comment]
@leshow you can use [this](https://docs.rs/built/0.3.0/built/util/fn.detect_ci.html) to detect CI.

**Comment by leshow** [comment]
Okay, I can certainly add that. Another thing that just came to mind: I haven't gated the progressbar behind any kind of `log::Level`, I could certainly add that also.

**Comment by atroche** [comment]
I'm fine with showing it by default, but happy to be talked out of it! Let me know if you need any help getting this to run on Travis.

**Comment by atroche** [comment]
@leshow last time it ran it seemed to be downloading the Chrome binary very slowly. We may want to cache the download, if we can, and if the download times out we may need to disable `fetcher` altogether on CI.

**Comment by leshow** [comment]
Okay, @atroche I've added `~/.local/share/` to the cached directories for travis

**Comment by leshow** [comment]
I'm not quite sure how to solve the travis error here. If I remove `stutter`, won't clippy fail on stable?

**Comment by atroche** [comment]
I had that problem too, but somehow it's gone away after reinstalling my toolchain via rustup. Pretty mysterious!

**Comment by atroche** [comment]
(stutter is fixed on the latest master now)

**Comment by leshow** [comment]
@atroche I was able to get the CI to pass finally on linux by looking at puppeteer's travis.yml. But one of the `.travis.yml` params required to make that happen causes the mac CI to fail.

On mac we'd need to pass in the `--no-sandbox` command *I think* in order to get this working, but we'd also need to remove that `sysctl` bit.

Sorry for all the churn on this branch.

**Comment by atroche** [comment]
No need to apologise, dude. I'm impressed with your tenacity.

**Comment by leshow** [comment]
Thanks. I've managed to fix the error I previously mentioned, but alas, I've no idea how to fix the current one. Looks like it's exploding during compiling some third party dep `security-framework`

**Comment by leshow** [comment]
@atroche do you have a machine with osx? Does this branch fail to compile for you?

**Comment by atroche** [comment]
Hmm, it compiles for me on Mac OS 10.14.3. Did you see the note in the Travis logs saying:

`Perhaps two different versions of crate `core_foundation_sys` are being used?`

Is there a way to tell if that's true?

**Comment by atroche** [comment]
Output of `cargo tree`:

```
headless_chrome v0.1.3 (/Users/alistair/Code/rust/rust-headless-chrome)
├── base64 v0.10.1
│   └── byteorder v1.3.1
├── chrono v0.4.6
│   ├── num-integer v0.1.39
│   │   └── num-traits v0.2.6
│   ├── num-traits v0.2.6 (*)
│   └── time v0.1.42
│       └── libc v0.2.49
│       [dev-dependencies]
│       └── winapi v0.3.6
├── derive_builder v0.7.1
│   ├── darling v0.8.5
│   │   ├── darling_core v0.8.5
│   │   │   ├── fnv v1.0.6
│   │   │   ├── ident_case v1.0.0
│   │   │   ├── proc-macro2 v0.4.27
│   │   │   │   └── unicode-xid v0.1.0
│   │   │   ├── quote v0.6.11
│   │   │   │   └── proc-macro2 v0.4.27 (*)
│   │   │   └── syn v0.15.26
│   │   │       ├── proc-macro2 v0.4.27 (*)
│   │   │       ├── quote v0.6.11 (*)
│   │   │       └── unicode-xid v0.1.0 (*)
│   │   └── darling_macro v0.8.5
│   │       ├── darling_core v0.8.5 (*)
│   │       ├── quote v0.6.11 (*)
│   │       └── syn v0.15.26 (*)
│   ├── derive_builder_core v0.4.1
│   │   ├── darling v0.8.5 (*)
│   │   ├── proc-macro2 v0.4.27 (*)
│   │   ├── quote v0.6.11 (*)
│   │   └── syn v0.15.26 (*)
│   ├── proc-macro2 v0.4.27 (*)
│   ├── quote v0.6.11 (*)
│   └── syn v0.15.26 (*)
├── directories v1.0.2
│   └── libc v0.2.49 (*)
├── env_logger v0.6.0
│   ├── atty v0.2.11
│   │   └── libc v0.2.49 (*)
│   ├── humantime v1.2.0
│   │   └── quick-error v1.2.2
│   ├── log v0.4.6
│   │   └── cfg-if v0.1.6
│   ├── regex v1.1.2
│   │   ├── aho-corasick v0.6.10
│   │   │   └── memchr v2.2.0
│   │   ├── memchr v2.2.0 (*)
│   │   ├── regex-syntax v0.6.5
│   │   │   └── ucd-util v0.1.3
│   │   ├── thread_local v0.3.6
│   │   │   └── lazy_static v1.3.0
│   │   └── utf8-ranges v1.0.2
│   └── termcolor v1.0.4
├── failure v0.1.5
│   ├── backtrace v0.3.14
│   │   ├── cfg-if v0.1.6 (*)
│   │   ├── libc v0.2.49 (*)
│   │   └── rustc-demangle v0.1.13
│   │   [build-dependencies]
│   │   └── autocfg v0.1.2
│   └── failure_derive v0.1.5
│       ├── proc-macro2 v0.4.27 (*)
│       ├── quote v0.6.11 (*)
│       ├── syn v0.15.26 (*)
│       └── synstructure v0.10.1
│           ├── proc-macro2 v0.4.27 (*)
│           ├── quote v0.6.11 (*)
│           ├── syn v0.15.26 (*)
│           └── unicode-xid v0.1.0 (*)
├── indicatif v0.11.0
│   ├── console v0.7.5
│   │   ├── atty v0.2.11 (*)
│   │   ├── clicolors-control v1.0.0
│   │   │   ├── lazy_static v1.3.0 (*)
│   │   │   └── libc v0.2.49 (*)
│   │   ├── lazy_static v1.3.0 (*)
│   │   ├── libc v0.2.49 (*)
│   │   ├── parking_lot v0.7.1
│   │   │   ├── lock_api v0.1.5
│   │   │   │   ├── owning_ref v0.4.0
│   │   │   │   │   └── stable_deref_trait v1.1.1
│   │   │   │   └── scopeguard v0.3.3
│   │   │   └── parking_lot_core v0.4.0
│   │   │       ├── libc v0.2.49 (*)
│   │   │       ├── rand v0.6.5
│   │   │       │   ├── libc v0.2.49 (*)
│   │   │       │   ├── rand_chacha v0.1.1
│   │   │       │   │   └── rand_core v0.3.1
│   │   │       │   │       └── rand_core v0.4.0
│   │   │       │   │   [build-dependencies]
│   │   │       │   │   └── autocfg v0.1.2 (*)
│   │   │       │   ├── rand_core v0.4.0 (*)
│   │   │       │   ├── rand_hc v0.1.0
│   │   │       │   │   └── rand_core v0.3.1 (*)
│   │   │       │   ├── rand_isaac v0.1.1
│   │   │       │   │   └── rand_core v0.3.1 (*)
│   │   │       │   ├── rand_jitter v0.1.3
│   │   │       │   │   ├── libc v0.2.49 (*)
│   │   │       │   │   └── rand_core v0.4.0 (*)
│   │   │       │   ├── rand_os v0.1.2
│   │   │       │   │   ├── libc v0.2.49 (*)
│   │   │       │   │   └── rand_core v0.4.0 (*)
│   │   │       │   ├── rand_pcg v0.1.2
│   │   │       │   │   └── rand_core v0.4.0 (*)
│   │   │       │   │   [build-dependencies]
│   │   │       │   │   └── autocfg v0.1.2 (*)
│   │   │       │   └── rand_xorshift v0.1.1
│   │   │       │       └── rand_core v0.3.1 (*)
│   │   │       │   [build-dependencies]
│   │   │       │   └── autocfg v0.1.2 (*)
│   │   │       └── smallvec v0.6.9
│   │   │       [build-dependencies]
│   │   │       └── rustc_version v0.2.3
│   │   │           └── semver v0.9.0
│   │   │               └── semver-parser v0.7.0
│   │   ├── regex v1.1.2 (*)
│   │   ├── termios v0.3.1
│   │   │   └── libc v0.2.49 (*)
│   │   └── unicode-width v0.1.5
│   ├── lazy_static v1.3.0 (*)
│   ├── number_prefix v0.2.8
│   │   └── num-traits v0.2.6 (*)
│   ├── parking_lot v0.7.1 (*)
│   └── regex v1.1.2 (*)
├── log v0.4.6 (*)
├── rand v0.6.5 (*)
├── regex v1.1.2 (*)
├── reqwest v0.9.10
│   ├── base64 v0.10.1 (*)
│   ├── bytes v0.4.11
│   │   ├── byteorder v1.3.1 (*)
│   │   └── iovec v0.1.2
│   │       └── libc v0.2.49 (*)
│   ├── encoding_rs v0.8.17
│   │   └── cfg-if v0.1.6 (*)
│   ├── futures v0.1.25
│   ├── http v0.1.16
│   │   ├── bytes v0.4.11 (*)
│   │   ├── fnv v1.0.6 (*)
│   │   └── itoa v0.4.3
│   ├── hyper v0.12.24
│   │   ├── bytes v0.4.11 (*)
│   │   ├── futures v0.1.25 (*)
│   │   ├── futures-cpupool v0.1.8
│   │   │   ├── futures v0.1.25 (*)
│   │   │   └── num_cpus v1.10.0
│   │   │       └── libc v0.2.49 (*)
│   │   ├── h2 v0.1.16
│   │   │   ├── byteorder v1.3.1 (*)
│   │   │   ├── bytes v0.4.11 (*)
│   │   │   ├── fnv v1.0.6 (*)
│   │   │   ├── futures v0.1.25 (*)
│   │   │   ├── http v0.1.16 (*)
│   │   │   ├── indexmap v1.0.2
│   │   │   ├── log v0.4.6 (*)
│   │   │   ├── slab v0.4.2
│   │   │   ├── string v0.1.3
│   │   │   └── tokio-io v0.1.11
│   │   │       ├── bytes v0.4.11 (*)
│   │   │       ├── futures v0.1.25 (*)
│   │   │       └── log v0.4.6 (*)
│   │   ├── http v0.1.16 (*)
│   │   ├── httparse v1.3.3
│   │   ├── iovec v0.1.2 (*)
│   │   ├── itoa v0.4.3 (*)
│   │   ├── log v0.4.6 (*)
│   │   ├── net2 v0.2.33
│   │   │   ├── cfg-if v0.1.6 (*)
│   │   │   └── libc v0.2.49 (*)
│   │   ├── time v0.1.42 (*)
│   │   ├── tokio v0.1.15
│   │   │   ├── bytes v0.4.11 (*)
│   │   │   ├── futures v0.1.25 (*)
│   │   │   ├── mio v0.6.16
│   │   │   │   ├── iovec v0.1.2 (*)
│   │   │   │   ├── lazycell v1.2.1
│   │   │   │   ├── libc v0.2.49 (*)
│   │   │   │   ├── log v0.4.6 (*)
│   │   │   │   ├── net2 v0.2.33 (*)
│   │   │   │   └── slab v0.4.2 (*)
│   │   │   ├── num_cpus v1.10.0 (*)
│   │   │   ├── tokio-current-thread v0.1.4
│   │   │   │   ├── futures v0.1.25 (*)
│   │   │   │   └── tokio-executor v0.1.6
│   │   │   │       ├── crossbeam-utils v0.6.5
│   │   │   │       │   ├── cfg-if v0.1.6 (*)
│   │   │   │       │   └── lazy_static v1.3.0 (*)
│   │   │   │       └── futures v0.1.25 (*)
│   │   │   ├── tokio-executor v0.1.6 (*)
│   │   │   ├── tokio-io v0.1.11 (*)
│   │   │   ├── tokio-reactor v0.1.8
│   │   │   │   ├── crossbeam-utils v0.6.5 (*)
│   │   │   │   ├── futures v0.1.25 (*)
│   │   │   │   ├── lazy_static v1.3.0 (*)
│   │   │   │   ├── log v0.4.6 (*)
│   │   │   │   ├── mio v0.6.16 (*)
│   │   │   │   ├── num_cpus v1.10.0 (*)
│   │   │   │   ├── parking_lot v0.7.1 (*)
│   │   │   │   ├── slab v0.4.2 (*)
│   │   │   │   ├── tokio-executor v0.1.6 (*)
│   │   │   │   └── tokio-io v0.1.11 (*)
│   │   │   │   [dev-dependencies]
│   │   │   │   └── num_cpus v1.10.0 (*)
│   │   │   ├── tokio-threadpool v0.1.11
│   │   │   │   ├── crossbeam v0.6.0
│   │   │   │   │   ├── cfg-if v0.1.6 (*)
│   │   │   │   │   ├── crossbeam-channel v0.3.8
│   │   │   │   │   │   ├── crossbeam-utils v0.6.5 (*)
│   │   │   │   │   │   └── smallvec v0.6.9 (*)
│   │   │   │   │   ├── crossbeam-deque v0.6.3
│   │   │   │   │   │   ├── crossbeam-epoch v0.7.1
│   │   │   │   │   │   │   ├── arrayvec v0.4.10
│   │   │   │   │   │   │   │   └── nodrop v0.1.13
│   │   │   │   │   │   │   ├── cfg-if v0.1.6 (*)
│   │   │   │   │   │   │   ├── crossbeam-utils v0.6.5 (*)
│   │   │   │   │   │   │   ├── lazy_static v1.3.0 (*)
│   │   │   │   │   │   │   ├── memoffset v0.2.1
│   │   │   │   │   │   │   └── scopeguard v0.3.3 (*)
│   │   │   │   │   │   └── crossbeam-utils v0.6.5 (*)
│   │   │   │   │   ├── crossbeam-epoch v0.7.1 (*)
│   │   │   │   │   ├── crossbeam-utils v0.6.5 (*)
│   │   │   │   │   ├── lazy_static v1.3.0 (*)
│   │   │   │   │   ├── num_cpus v1.10.0 (*)
│   │   │   │   │   └── parking_lot v0.7.1 (*)
│   │   │   │   ├── crossbeam-channel v0.3.8 (*)
│   │   │   │   ├── crossbeam-deque v0.6.3 (*)
│   │   │   │   ├── crossbeam-utils v0.6.5 (*)
│   │   │   │   ├── futures v0.1.25 (*)
│   │   │   │   ├── log v0.4.6 (*)
│   │   │   │   ├── num_cpus v1.10.0 (*)
│   │   │   │   ├── rand v0.6.5 (*)
│   │   │   │   ├── slab v0.4.2 (*)
│   │   │   │   └── tokio-executor v0.1.6 (*)
│   │   │   └── tokio-timer v0.2.10
│   │   │       ├── crossbeam-utils v0.6.5 (*)
│   │   │       ├── futures v0.1.25 (*)
│   │   │       ├── slab v0.4.2 (*)
│   │   │       └── tokio-executor v0.1.6 (*)
│   │   │   [dev-dependencies]
│   │   │   └── num_cpus v1.10.0 (*)
│   │   ├── tokio-executor v0.1.6 (*)
│   │   ├── tokio-io v0.1.11 (*)
│   │   ├── tokio-reactor v0.1.8 (*)
│   │   ├── tokio-tcp v0.1.3
│   │   │   ├── bytes v0.4.11 (*)
│   │   │   ├── futures v0.1.25 (*)
│   │   │   ├── iovec v0.1.2 (*)
│   │   │   ├── mio v0.6.16 (*)
│   │   │   ├── tokio-io v0.1.11 (*)
│   │   │   └── tokio-reactor v0.1.8 (*)
│   │   ├── tokio-threadpool v0.1.11 (*)
│   │   ├── tokio-timer v0.2.10 (*)
│   │   └── want v0.0.6
│   │       ├── futures v0.1.25 (*)
│   │       ├── log v0.4.6 (*)
│   │       └── try-lock v0.2.2
│   ├── hyper-tls v0.3.1
│   │   ├── bytes v0.4.11 (*)
│   │   ├── futures v0.1.25 (*)
│   │   ├── hyper v0.12.24 (*)
│   │   ├── native-tls v0.2.2
│   │   │   ├── lazy_static v1.3.0 (*)
│   │   │   ├── libc v0.2.49 (*)
│   │   │   ├── security-framework v0.2.2
│   │   │   │   ├── core-foundation v0.5.1
│   │   │   │   │   ├── core-foundation-sys v0.5.1
│   │   │   │   │   │   └── libc v0.2.49 (*)
│   │   │   │   │   └── libc v0.2.49 (*)
│   │   │   │   ├── core-foundation-sys v0.5.1 (*)
│   │   │   │   ├── libc v0.2.49 (*)
│   │   │   │   └── security-framework-sys v0.2.3
│   │   │   │       ├── MacTypes-sys v2.1.0
│   │   │   │       │   └── libc v0.2.49 (*)
│   │   │   │       ├── core-foundation-sys v0.5.1 (*)
│   │   │   │       └── libc v0.2.49 (*)
│   │   │   ├── security-framework-sys v0.2.3 (*)
│   │   │   └── tempfile v3.0.7
│   │   │       ├── cfg-if v0.1.6 (*)
│   │   │       ├── libc v0.2.49 (*)
│   │   │       ├── rand v0.6.5 (*)
│   │   │       └── remove_dir_all v0.5.1
│   │   └── tokio-io v0.1.11 (*)
│   ├── libflate v0.1.20
│   │   ├── adler32 v1.0.3
│   │   ├── byteorder v1.3.1 (*)
│   │   └── crc32fast v1.1.2
│   │       └── cfg-if v0.1.6 (*)
│   ├── log v0.4.6 (*)
│   ├── mime v0.3.13
│   │   └── unicase v2.2.0
│   │       [build-dependencies]
│   │       └── version_check v0.1.5
│   ├── mime_guess v2.0.0-alpha.6
│   │   ├── mime v0.3.13 (*)
│   │   ├── phf v0.7.24
│   │   │   └── phf_shared v0.7.24
│   │   │       ├── siphasher v0.2.3
│   │   │       └── unicase v1.4.2
│   │   │           [build-dependencies]
│   │   │           └── version_check v0.1.5 (*)
│   │   └── unicase v1.4.2 (*)
│   │   [build-dependencies]
│   │   ├── phf_codegen v0.7.24
│   │   │   ├── phf_generator v0.7.24
│   │   │   │   ├── phf_shared v0.7.24 (*)
│   │   │   │   └── rand v0.6.5 (*)
│   │   │   └── phf_shared v0.7.24 (*)
│   │   └── unicase v1.4.2 (*)
│   ├── native-tls v0.2.2 (*)
│   ├── serde v1.0.88
│   │   └── serde_derive v1.0.88
│   │       ├── proc-macro2 v0.4.27 (*)
│   │       ├── quote v0.6.11 (*)
│   │       └── syn v0.15.26 (*)
│   │   [dev-dependencies]
│   │   └── serde_derive v1.0.88 (*)
│   ├── serde_json v1.0.38
│   │   ├── itoa v0.4.3 (*)
│   │   ├── ryu v0.2.7
│   │   └── serde v1.0.88 (*)
│   ├── serde_urlencoded v0.5.4
│   │   ├── dtoa v0.4.3
│   │   ├── itoa v0.4.3 (*)
│   │   ├── serde v1.0.88 (*)
│   │   └── url v1.7.2
│   │       ├── idna v0.1.5
│   │       │   ├── matches v0.1.8
│   │       │   ├── unicode-bidi v0.3.4
│   │       │   │   └── matches v0.1.8 (*)
│   │       │   └── unicode-normalization v0.1.8
│   │       │       └── smallvec v0.6.9 (*)
│   │       ├── matches v0.1.8 (*)
│   │       └── percent-encoding v1.0.1
│   ├── tokio v0.1.15 (*)
│   ├── tokio-executor v0.1.6 (*)
│   ├── tokio-io v0.1.11 (*)
│   ├── tokio-threadpool v0.1.11 (*)
│   ├── tokio-timer v0.2.10 (*)
│   ├── url v1.7.2 (*)
│   └── uuid v0.7.2
│       └── rand v0.6.5 (*)
├── serde v1.0.88 (*)
├── serde_derive v1.0.88 (*)
├── serde_json v1.0.38 (*)
├── tempfile v3.0.7 (*)
├── toml v0.4.10
│   └── serde v1.0.88 (*)
├── websocket v0.22.3
│   ├── base64 v0.10.1 (*)
│   ├── bitflags v1.0.4
│   ├── byteorder v1.3.1 (*)
│   ├── hyper v0.10.15
│   │   ├── base64 v0.9.3
│   │   │   ├── byteorder v1.3.1 (*)
│   │   │   └── safemem v0.3.0
│   │   ├── httparse v1.3.3 (*)
│   │   ├── language-tags v0.2.2
│   │   ├── log v0.3.9
│   │   │   └── log v0.4.6 (*)
│   │   ├── mime v0.2.6
│   │   │   └── log v0.3.9 (*)
│   │   ├── num_cpus v1.10.0 (*)
│   │   ├── time v0.1.42 (*)
│   │   ├── traitobject v0.1.0
│   │   ├── typeable v0.1.2
│   │   ├── unicase v1.4.2 (*)
│   │   └── url v1.7.2 (*)
│   ├── rand v0.6.5 (*)
│   ├── sha1 v0.6.0
│   ├── unicase v1.4.2 (*)
│   └── url v1.7.2 (*)
└── zip v0.5.0
    ├── bzip2 v0.3.3
    │   ├── bzip2-sys v0.1.7
    │   │   └── libc v0.2.49 (*)
    │   │   [build-dependencies]
    │   │   └── cc v1.0.29
    │   └── libc v0.2.49 (*)
    ├── crc32fast v1.1.2 (*)
    ├── libflate v0.1.20 (*)
    ├── podio v0.1.6
    └── time v0.1.42 (*)
[dev-dependencies]
├── jpeg-decoder v0.1.15
│   └── byteorder v1.3.1 (*)
├── png v0.14.0
│   ├── bitflags v1.0.4 (*)
│   ├── deflate v0.7.19
│   │   ├── adler32 v1.0.3 (*)
│   │   └── byteorder v1.3.1 (*)
│   ├── inflate v0.4.5
│   │   └── adler32 v1.0.3 (*)
│   └── num-iter v0.1.37
│       ├── num-integer v0.1.39 (*)
│       └── num-traits v0.2.6 (*)
└── tiny_http v0.6.2
    ├── ascii v0.8.7
    ├── chrono v0.4.6 (*)
    ├── chunked_transfer v0.3.1
    ├── log v0.4.6 (*)
    └── url v1.7.2 (*)
```

**Comment by atroche** [comment]
Hmm, I'll try pushing a branch with `osx_image: xcode10.1` set.

**Comment by atroche** [comment]
(https://docs.travis-ci.com/user/reference/osx/)

**Comment by atroche** [comment]
Let's see how that goes: https://travis-ci.com/atroche/rust-headless-chrome/jobs/181940046

**Comment by atroche** [comment]
So I'm noticing that when running `cargo test` multiple threads will try to download Chromium at once. Perhaps we need a lock of some kind?

**Comment by leshow** [comment]
If that was the case wouldn't the linux CI fail also?

**Comment by leshow** [comment]
... and it should also cache the download. all of `~/.local/share` is cached.

**Comment by leshow** [comment]
OH! perhaps we need to add the OSX dl dir to the travis cache

**Comment by atroche** [comment]
@leshow downloading / extracting many at once might not be a failure, but it's probably not a great idea. it also might partly explain why extracting takes so long at the moment.

**Comment by leshow** [comment]
That's odd, I haven't seen that before. In any case it's not occurring after the changes I made. OSX is erroring out now on the sandbox stuff now. 

This is because the `sysctl` command required to run on linux can't be set on mac. I don't know if there is an equivalent option. From puppeteer's docs it doesn't look like they support mac, as a workaround we could send the --no-sandbox for the unit tests, but there's currently no way to pass the option to `LaunchOptions`. 

As a temporary workaround, I can bring back `default_executable` and use the system chrome for the tests. I think removing this and passing no-sandbox to the tests after we have launch params is the right fix here most likely.

**Comment by leshow** [comment]
I've added back default_executable and am using that for all the tests, so `Fetcher` won't run in the CI anymore. It fails on a browser test `test_extension`, which is somehow calling the mac download URL?

**Comment by leshow** [comment]
Well, that took a while. @atroche I'm skipping `Fetcher` for all the tests and examples.

**Comment by atroche** [comment]
I think we still need to deal with the multiple downloads / extractions at once thing, and also make it clear in the docs what's going to happen unless they specify a `path`.

**Comment by leshow** [comment]
I don't follow, multiple downloads and extractions won't happen. I beleive that was because tests are run in parallel and multiple tests instantiated `LaunchOptions`, which in turn instantiated `Fetcher`. Specifying a path skips `Fetcher` so no download/extract happens.

On a users machine, they would have to execute multiple `Process` in parallel without ever having run it before (otherwise the download will skip). Unless I'm missing something. Another weird thing: I never observed that multiple download behaviour on my machine or in the CI on linux as best I can remember, just on OSX.

In order to use a downloaded chrome, but not download many instances during `cargo test`. I was able to put:

```

**Comment by atroche** [comment]
I suppose it's unlikely that users will instantiate multiple `Process`es on their very first run, and even then if they do it's not the end of the world. I'd be happy to merge this, just let me use that GitHub 'suggest changes' on the docstring for Browser so people realise that downloading is the default unless they specify a path.

**Comment by atroche** [comment]
Ah, okay, couldn't suggest changes to that part of the PR, but how's this?

```
/// With the default [LaunchOptions](../process/LaunchOptions.struct.html), we will automatically
/// download a known-good revision of Chromium into your $XDG_DATA_DIR. You can specify your own
/// path to a binary, or make use of the `default_executable` function to use your already-installed
/// copy of Chrome.
```

**Comment by atroche** [comment]
Oh, I don't suppose there's an easy way to not have to pull in `tokio`?

**Comment by leshow** [comment]
@atroche I think `reqwest` may be pulling that in. I'll take a look and see if we can pass a compile feature to only use sync IO. 

I'm happy to put more improvements to this in another PR in the future, something that would fail or print a message if another download is currently running. Another thing we could do is attempt a download in `build.rs` also, that way when headless_chrome runs the first time there's likely already a version available.

**Comment by leshow** [comment]
I added the documentation. I didn't find any way to not use tokio with reqwest. Even with raw hyper it looks like it's still a hard dependency.

**Comment by atroche** [comment]
problem with CI was result of the wait helpers refactor PR, have fixed and merged. thanks so much @leshow for persisting with this!

**Comment by leshow** [comment]
Thank you for accepting!

**Pull request merged**